src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceMarginViewMargin.xaml.cs

```diff
@@ -0,0 +1,244 @@
+        private readonly IUIThreadOperationExecutor _operationExecutor;
+        private readonly IOptionService _optionService;
+        private readonly string _languageName;
+        private Dictionary<SnapshotSpan, MarginGlyph.InheritanceMargin> _snapshotSpanToMargin;
+        // Same size as the glyph margin
+        private const double HeightAndWidthOfMargin = 18;
+
+        public InheritanceMarginViewMargin(
+            IWpfTextViewHost textViewHost,
+            IThreadingContext threadingContext,
+            IStreamingFindUsagesPresenter streamingFindUsagesPresenter,
+            IUIThreadOperationExecutor operationExecutor,
+            IClassificationFormatMap classificationFormatMap,
+            ClassificationTypeMap classificationTypeMap,
+            ITagAggregator<InheritanceMarginTag> tagAggregator,
+            Document document)
+        {
+            InitializeComponent();
+            _textViewHost = textViewHost;
+            _threadingContext = threadingContext;
+            _streamingFindUsagesPresenter = streamingFindUsagesPresenter;
+            _classificationFormatMap = classificationFormatMap;
+            _classificationTypeMap = classificationTypeMap;
+            _operationExecutor = operationExecutor;
+            _textView = textViewHost.TextView;
+            _tagAggregator = tagAggregator;
+            _snapshotSpanToMargin = new Dictionary<SnapshotSpan, MarginGlyph.InheritanceMargin>();
+
+            _tagAggregator.BatchedTagsChanged += OnTagsChanged;
+            _textView.LayoutChanged += OnLayoutChanged;
+            _textView.ZoomLevelChanged += OnZoomLevelChanged;
+
+            MainCanvas.Width = MarginSize;
+
+            MainCanvas.LayoutTransform = new ScaleTransform(
+                scaleX: _textView.ZoomLevel / 100,
+                scaleY: _textView.ZoomLevel / 100);
+            MainCanvas.LayoutTransform.Freeze();
+
+            _optionService = document.Project.Solution.Workspace.Services.GetRequiredService<IOptionService>();
+            _languageName = document.Project.Language;
+        }
+
+        private void OnZoomLevelChanged(object sender, ZoomLevelChangedEventArgs e)
+        {
+            // Set ZoomTransform for Canvas.
+            // Don't need to update the ZoomLevel for glpyhs here, when zoom level changes, OnLayoutChanged will be
+            // fired all the margin will be recreated.
+            MainCanvas.LayoutTransform = e.ZoomTransform;
+        }
+
+        private void OnLayoutChanged(object sender, TextViewLayoutChangedEventArgs e)
+        {
+            SetSnapshotAndUpdate(e.NewOrReformattedLines);
+            foreach (var line in e.NewOrReformattedLines)
+            {
+                RemoveGlyphByVisualSpan(line.Extent);
+                RefreshGlyphsOver(line);
+            }
+        }
+
+        private void SetSnapshotAndUpdate(IReadOnlyCollection<ITextViewLine> newOrReformattedLines)
```

Comment:
Code here is mostly based on http://index/?leftProject=Microsoft.VisualStudio.Platform.VSEditor&leftSymbol=menkdnpzodoc&file=GlyphMargin.cs&line=26 the editor side code. (Simplify it a little bit because our scenario is much easier)